Source/WebCore/animation/ScrollTimeline.cpp

```diff
@@ -138,4 +140,40 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentW
     return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::No;
 }
 
+ScrollableArea* ScrollTimeline::scrollableAreaForSourceRenderer(RenderElement* renderer, Ref<Document> document)
+{
+    CheckedPtr renderBox = dynamicDowncast<RenderBox>(renderer);
+    if (!renderBox)
+        return nullptr;
+
+    if (renderer->element() == document->documentElement())
+        return &renderer->view().frameView();
+    return (renderBox->canBeScrolledAndHasScrollableArea() && renderBox->layer()) ? renderBox->layer()->scrollableArea() : nullptr;
+}
+
+ScrollTimeline::Data ScrollTimeline::computeScrollTimelineData() const
+{
+    if (!m_source)
+        return { };
+
+    auto* sourceScrollableArea = scrollableAreaForSourceRenderer(m_source->renderer(), m_source->document());
+    if (!sourceScrollableArea)
+        return { };
+
+    // https://drafts.csswg.org/scroll-animations-1/#scroll-timeline-progress
+    // Progress (the current time) for a scroll progress timeline is calculated as:
+    // scroll offset ÷ (scrollable overflow size − scroll container size)
+
+    float maxScrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->maximumScrollOffset().y() : sourceScrollableArea->maximumScrollOffset().x();
+    float scrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->scrollPosition().y() : sourceScrollableArea->scrollPosition().x();
```

Comment:
Beware of `scrollPosition` vs `scrollOffset`, and don't mix and match. They differ in RTL horizontal scrollers, so you need some tests for those. But use `scrollOffset()` here.

Reply:
Fixed.